File: weil_code_generator.py
import scipy.io as sio

def weil_code_generator(N, N0, w, p):
    residue = []
    for i in range(int((N + 1) / 2)):
        res = i ** 2 % N
        if res not in residue:
            residue.append(res)
    legendre = [0] * N
    for i in range(1, N):
        if i in residue:
            legendre[i] = 1
    weil_code = [0] * N
    for i in range(N):
        weil_code[i] = (legendre[i] + legendre[(i + w) % N]) % 2
    weil_out = []
    # for i in range(N0):
    
    #     weil_out = weil_out + 5*[weil_code[(p - 1 + i)]]
    if (N0 + p - 2) < N:
        weil_out = weil_code[p - 1: p - 1 + N0]
    else:
        weil_out = weil_code[p - 1:] + weil_code[:(p - 1 + N0) % N]
    return weil_out

r = weil_code_generator(10243, 10230, 4318, 341)
s = weil_code_generator(3607, 1800, 269, 1889)
t = weil_code_generator(10243, 10230, 796, 7575)
# print(r[0:24])
# print(r[-24:])
# print(s[0:24])
# print(s[-24:])
sio.savemat('weil.mat', {'data':r, 'child':s, 'pilot':t})

File: weil_code_generator.m
% generate B1C code 

function weil_out = weil_code_generator(N, N0, w, p)
    residu = [];
    for i = 0 : (N - 1) / 2
        res = mod(i^2, N);
        if ~ismember(res, residu)
            residu = [residu res];
        end
    end
    i = 0:N - 1;
    l = ismember(i, residu);
    weil_code = zeros(1, N);
    weil_code(1:N - w) = mod(l(1:N - w) + l(w + 1:N), 2);
    if w ~= 0
        weil_code(N - w + 1:N) = mod(l(N - w + 1:N) + l(1:w), 2);
    end
    if N0 + p - 1 < N
        weil_out = weil_code(p:p + N0 - 1);
    else
        weil_out = [weil_code(p:end) weil_code(1:mod(p+N0 - 1, N))];
    end
    z = (weil_out == 0);
    o = (weil_out == 1);
    weil_out(z) = 1;
    weil_out(o) = -1;
end
